StackOverflowError after rebuild Cluster N times

### Search before asking

- [x] I searched in the [issues](https://github.com/alibaba/fluss/issues) and found nothing similar.


### Fluss version

0.6.0 (latest release)

### Please describe the bug 🐞

The Flink tests hangs in https://github.com/alibaba/fluss/actions/runs/14175965698?pr=693. After investigating the maven logs, I found a lot of `StackOverflowError`. The reason is the value of `availableLocationsByPath` is nested under `Collections.unmodifiableList` for N levels. When iterating the values of   `availableLocationsByPath`, there is a StackOverflowError thrown. See https://stackoverflow.com/questions/29026487/stackoverflowerror-collectionsunmodifiablelist


```
01:02:31,479 [fluss-write-sender-thread-1] ERROR com.alibaba.fluss.client.write.Sender                        [] - Uncaught error in Fluss write sender thread: 
com.alibaba.fluss.exception.FlussRuntimeException: Failed to update metadata
	at com.alibaba.fluss.client.metadata.MetadataUpdater.updateMetadata(MetadataUpdater.java:268) ~[fluss-client-0.7-SNAPSHOT.jar:0.7-SNAPSHOT]
	at com.alibaba.fluss.client.metadata.MetadataUpdater.updatePhysicalTableMetadata(MetadataUpdater.java:245) ~[fluss-client-0.7-SNAPSHOT.jar:0.7-SNAPSHOT]
	at com.alibaba.fluss.client.write.Sender.sendWriteData(Sender.java:198) ~[fluss-client-0.7-SNAPSHOT.jar:0.7-SNAPSHOT]
	at com.alibaba.fluss.client.write.Sender.runOnce(Sender.java:179) ~[fluss-client-0.7-SNAPSHOT.jar:0.7-SNAPSHOT]
	at com.alibaba.fluss.client.write.Sender.run(Sender.java:147) [fluss-client-0.7-SNAPSHOT.jar:0.7-SNAPSHOT]
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_432]
	at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_432]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_432]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_432]
	at java.lang.Thread.run(Thread.java:750) [?:1.8.0_432]
Caused by: java.lang.StackOverflowError
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection.iterator(Collections.java:1040) ~[?:1.8.0_432]
	at java.util.Collections$UnmodifiableCollection$1.<init>(Collections.java:1041) ~[?:1.8.0_432]
```

### Solution

_No response_

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

--- a/fluss-client/src/main/java/com/alibaba/fluss/client/metadata/MetadataUpdater.java
+++ b/fluss-client/src/main/java/com/alibaba/fluss/client/metadata/MetadataUpdater.java
@@ -278,43 +278,37 @@ public class MetadataUpdater {
         List<InetSocketAddress> inetSocketAddresses =
                 ClientUtils.parseAndValidateAddresses(conf.get(ConfigOptions.BOOTSTRAP_SERVERS));
         Cluster cluster = null;
+        Exception exception = null;
         for (InetSocketAddress address : inetSocketAddresses) {
-            cluster = tryToInitializeCluster(rpcClient, address);
-            if (cluster != null) {
+            try {
+                cluster = tryToInitializeCluster(rpcClient, address);
                 break;
+            } catch (Exception e) {
+                exception = e;
             }
         }
 
-        if (cluster == null) {
+        if (cluster == null && exception != null) {
             String errorMsg =
                     "Failed to initialize fluss client connection to server because no "
                             + "bootstrap server is validate. bootstrap servers: "
                             + inetSocketAddresses;
             LOG.error(errorMsg);
-            throw new IllegalStateException(errorMsg);
+            throw new IllegalStateException(errorMsg, exception);
         }
 
         return cluster;
     }
 
-    private static @Nullable Cluster tryToInitializeCluster(
-            RpcClient rpcClient, InetSocketAddress address) {
+    private static Cluster tryToInitializeCluster(RpcClient rpcClient, InetSocketAddress address)
+            throws Exception {
         ServerNode serverNode =
                 new ServerNode(
                         -1, address.getHostString(), address.getPort(), ServerType.COORDINATOR);
-        try {
-            AdminReadOnlyGateway adminReadOnlyGateway =
-                    GatewayClientProxy.createGatewayProxy(
-                            () -> serverNode, rpcClient, AdminReadOnlyGateway.class);
-            return sendMetadataRequestAndRebuildCluster(
-                    adminReadOnlyGateway, Collections.emptySet());
-        } catch (Exception e) {
-            LOG.error(
-                    "Failed to initialize fluss client connection to bootstrap server: {}",
-                    address,
-                    e);
-            return null;
-        }
+        AdminReadOnlyGateway adminReadOnlyGateway =
+                GatewayClientProxy.createGatewayProxy(
+                        () -> serverNode, rpcClient, AdminReadOnlyGateway.class);
+        return sendMetadataRequestAndRebuildCluster(adminReadOnlyGateway, Collections.emptySet());
     }
 
     /** Invalid the bucket metadata for the given physical table paths. */

--- a/fluss-client/src/test/java/com/alibaba/fluss/client/metadata/MetadataUpdaterTest.java
+++ b/fluss-client/src/test/java/com/alibaba/fluss/client/metadata/MetadataUpdaterTest.java
@@ -16,8 +16,14 @@
 
 package com.alibaba.fluss.client.metadata;
 
+import com.alibaba.fluss.client.Connection;
+import com.alibaba.fluss.client.ConnectionFactory;
+import com.alibaba.fluss.client.admin.Admin;
+import com.alibaba.fluss.client.utils.MetadataUtils;
 import com.alibaba.fluss.cluster.Cluster;
 import com.alibaba.fluss.cluster.ServerNode;
+import com.alibaba.fluss.config.Configuration;
+import com.alibaba.fluss.metadata.TablePath;
 import com.alibaba.fluss.rpc.RpcClient;
 import com.alibaba.fluss.server.testutils.FlussClusterExtension;
 
@@ -27,6 +33,7 @@ import org.junit.jupiter.api.extension.RegisterExtension;
 import java.util.Collections;
 import java.util.List;
 
+import static com.alibaba.fluss.record.TestData.DATA1_TABLE_DESCRIPTOR;
 import static org.assertj.core.api.Assertions.assertThat;
 
 /** Test for update metadata of {@link MetadataUpdater}. */
@@ -35,6 +42,36 @@ class MetadataUpdaterTest {
     @RegisterExtension
     public static final FlussClusterExtension FLUSS_CLUSTER_EXTENSION =
             FlussClusterExtension.builder().setNumOfTabletServers(2).build();
+
+    @Test
+    void testRebuildClusterNTimes() throws Exception {
+        Configuration clientConf = FLUSS_CLUSTER_EXTENSION.getClientConfig();
+        Connection conn = ConnectionFactory.createConnection(clientConf);
+        Admin admin = conn.getAdmin();
+        TablePath tablePath = TablePath.of("fluss", "test");
+        admin.createTable(tablePath, DATA1_TABLE_DESCRIPTOR, true).get();
+        admin.close();
+        conn.close();
+
+        RpcClient rpcClient = FLUSS_CLUSTER_EXTENSION.getRpcClient();
+        MetadataUpdater metadataUpdater = new MetadataUpdater(clientConf, rpcClient);
+        // update metadata
+        metadataUpdater.updateMetadata(Collections.singleton(tablePath), null, null);
+        Cluster cluster = metadataUpdater.getCluster();
+
+        // repeat 20K times to reproduce StackOverflowError if there is
+        // any N levels UnmodifiableCollection
+        for (int i = 0; i < 20000; i++) {
+            cluster =
+                    MetadataUtils.sendMetadataRequestAndRebuildCluster(
+                            FLUSS_CLUSTER_EXTENSION.newCoordinatorClient(),
+                            true,
+                            cluster,
+                            null,
+                            null,
+                            null);
+        }
+    }
 
     @Test
     void testUpdateWithEmptyMetadataResponse() throws Exception {

--- a/fluss-client/src/test/java/com/alibaba/fluss/client/metadata/MetadataUpdaterTest.java
+++ b/fluss-client/src/test/java/com/alibaba/fluss/client/metadata/MetadataUpdaterTest.java
@@ -116,5 +116,8 @@ class MetadataUpdaterTest {
         assertThat(metadataUpdater.getCluster().getAliveTabletServers())
                 .isEqualTo(newCluster.getAliveTabletServers())
                 .hasSize(1);
+
+        // recover the coordinator
+        FLUSS_CLUSTER_EXTENSION.startCoordinatorServer();
     }
 }

--- a/fluss-common/src/main/java/com/alibaba/fluss/cluster/Cluster.java
+++ b/fluss-common/src/main/java/com/alibaba/fluss/cluster/Cluster.java
@@ -80,7 +80,9 @@ public final class Cluster {
         for (Map.Entry<PhysicalTablePath, List<BucketLocation>> entry :
                 bucketLocationsByPath.entrySet()) {
             PhysicalTablePath physicalTablePath = entry.getKey();
-            List<BucketLocation> bucketsForTable = entry.getValue();
+            // avoid StackOverflowError on N levels UnmodifiableCollection,
+            // see https://stackoverflow.com/a/29027474
+            List<BucketLocation> bucketsForTable = new ArrayList<>(entry.getValue());
             // Optimise for the common case where all buckets are available.
             boolean foundUnavailableBucket = false;
             List<BucketLocation> availableBucketsForTable = new ArrayList<>(bucketsForTable.size());

